platform/utm_mgrs_utils.cpp

```diff
@@ -0,0 +1,567 @@
+
+  longitude = mod_angle(longitude + degToRad(zone_number_to_central_longitude(zone_number)));
+
+  lat = radToDeg(latitude);
+  lon = radToDeg(longitude);
+
+  return true;
+}
+
+// Convert MGRS parameters to UTM parameters and then use UTM to lat,lon conversion.
+bool MGRStoLatLon(double easting, double northing, int zone_code, char zone_letter, char square_code[2], double &lat, double &lon)
+{
+  // Convert easting and northing according to zone_code and square_code
+  if (zone_code < 1 || zone_code > 60)
+    return false;
+
+  if (zone_letter <= 'B' || zone_letter >= 'Y' || zone_letter == 'I' || zone_letter == 'O')
+    return false;
+
+  int set = zone_to_100k(zone_code);
+
+  char char1 = square_code[0];
+  char char2 = square_code[1];
+
+  if (char1 < 'A' || char2 < 'A' || char1 > 'Z' || char2 > 'Z' || char1 == 'I' || char2 == 'I' || char1 == 'O' || char2 == 'O')
+    return false;
+
+  float east100k = square_char_to_easting(char1, set);
+  if (east100k < 0)
+    return false;
+
+  float north100k = square_char_to_northing(char2, set);
+  if (north100k < 0)
+    return false;
+
+  double minNorthing = zone_to_min_northing(zone_letter);
+  if (minNorthing < 0)
+    return false;
+
+  while (north100k < minNorthing)
+    north100k += 2000000.0;
+
+  easting  += east100k;
+  northing += north100k;
+
+  return UTMtoLatLon(easting, northing, zone_code, zone_letter, lat, lon);
+}
+
+/**
+ * Given the first letter from a two-letter MGRS 100k zone, and given the
+ * MGRS table set for the zone number, figure out the easting value that
+ * should be added to the other, secondary easting value.
+ */
+double square_char_to_easting(char e, int set) {
+  int curCol = SET_ORIGIN_COLUMN_LETTERS[set - 1];
+  double eastingValue = 100000.0;
+  bool rewindMarker = false;
+
+  while (curCol != e)
+  {
+    curCol++;
+    if (curCol == 'I')
+      curCol++;
+    if (curCol == 'O')
+      curCol++;
+    if (curCol > 'Z')
+    {
+      if (rewindMarker)
+        return -1;
+      curCol = 'A';
+      rewindMarker = true;
+    }
+    eastingValue += 100000.0;
+  }
+
+  return eastingValue;
+}
+
+/**
+ * Given the second letter from a two-letter MGRS 100k zone, and given the
+ * MGRS table set for the zone number, figure out the northing value that
+ * should be added to the other, secondary northing value. You have to
+ * remember that Northings are determined from the equator, and the vertical
+ * cycle of letters mean a 2000000 additional northing meters. This happens
+ * approx. every 18 degrees of latitude. This method does *NOT* count any
+ * additional northings. You have to figure out how many 2000000 meters need
+ * to be added for the zone letter of the MGRS coordinate.
+ *
+ * @param n
+ *      second letter of the MGRS 100k zone
+ * @param set
+ *      the MGRS table set number, which is dependent on the UTM zone
+ *      number.
+ */
+double square_char_to_northing(char n, int set) {
+
+  if (n > 'V')
+    return -1;
```

Comment:
ditto

Reply:
Or if you find std::optional too heavy, please introduce kInvalid... constant instead.